### DISCUSSION TITLE
What is our stance on assignment expressions?

### DISCUSSION THREAD

**Discussion by adeak** [pyvista/pyvista#6232] [General]
@pyvista/developers I'd like to gauge how you feel about [assignment expressions](https://peps.python.org/pep-0572/), a fairly divisive feature. There are now a handful of instances using it in the codebase, [with a few worst offenders reverted](https://github.com/pyvista/pyvista/pull/6121) after I started picking at this in slack. I have no intention of blaming or shaming anyone who likes and uses the feature, I just want to see if there's any rough consensus among us about how to approach use of the feature. Ultimately we might want to amend [our contributing docs](https://docs.pyvista.org/version/stable/extras/developer_notes.html) to either explicitly encourage or restrict use of the feature.

I don't expect to come to a hasty conclusion here, especially since many stakeholders are not very active these days (myself included). Just trying to get the ball rolling to see where we stand :)

## Some background

For the record I've been an extreme hater of the feature even from before it was decided to be added to the language. I believe that the way Guido forced the feature through despite considerable pushback even from many core devs, and his subsequent resignation as BDFL out of frustration, is the OSS equivalent of a murder-suicide. But variety is the spice of life and I'm aware that most developers will be much more forgiving than me (which is great!). Please read this post with "in my opinion" generously added to each subjective statement, and apply the appropriate amount of salt.

My issue with assignment expressions (the "walrus operator", just to add some SEO to this thread) is that they almost guarantee spaghetti code except in the simplest use cases. In most cases where people are inclined to use the feature it leads to non-linear flow of execution, which adds disproportionate cognitive overhead to reading the code. I like to quote some horrendous examples from the PEP itself (which should show flagship cases for why we need the feature):

```python
# Compute partial sums in a list comprehension
total = 0
partial_sums = [total := total + v for v in values]
print("Total:", total)
```

This example is introduced as "_it allows a compact way of updating mutable state from a comprehension_". The problem here:
  1. It leaks state from inside a list comprehension, something that was "fixed" in Python 3.0 compared to 2.x.
  2. What this is doing really shouldn't be a list comprehension. Comprehensions are a functional programming pattern, and should have the same semantics as `map()`: take an iterable and do something with each value based on that value alone. If you start coupling to earlier iterations of the loop you really should use a proper loop, it is much much easier to read and understand the code's intent.

[The worst example](https://peps.python.org/pep-0572/#simplifying-list-comprehensions):

```python
results = [(x, y, x/y) for x in input_data if (y := f(x)) > 0]
```
This is just tragic. Although it's already a fact in comprehensions that the values are evaluated in the order of
```
[(3. derived value) (1. for variable in iterable) (2. conditions)]
```
the way the definition of the names at the start of the comprehension are bound at the end of the comprehension makes this very hard to read. Just use a loop:
```python
results = []
for x in input_data:
    y = f(x)
    if y > 0:
        results.append((x, y, x/y))
```
Yes, it's boring, but it's clear.

## Some pyvista history

I've looked up if we discussed this before anywhere, and [only found one PR thread from 2 years ago where I brought this up](https://github.com/pyvista/pyvista/pull/1716#issuecomment-998983753). The very inconclusive discussion was

> **adeak**: assignment expressions a.k.a the walrus operator: please, no, no no no no
> **akaszynski**: Agreed. I can see some edge cases where this is useful, but I don't encourage its use.
> **adeak**: I can pretty much only accept their use in the opening line of a loop or conditional clause, [minor elaboration]

Pointing this out as some "prior art", but opinions can also change.

## An explicit recommendation

Considering all of the above, my inclination would be to restrict or discourage the use of assignment expressions in the codebase. As a rule of thumb I tend to believe that the feature is only (mostly) harmless when put on the first line of conditionals or loops (but never inside comprehensions or conditional expressions). A typical example is a regex match. I'd still personally try to pre-define the variable before the loop/conditional in these cases, but sometimes (e.g. with multiple conditional branches) that is not feasible. For the record currently we have 16 instances of assignment expressions in `main`, out of which 13 are indeed in loop/conditional header lines.

So what do y'all think about the feature? We can also _encourage_ the feature if that ends up being the overwhelming feedback, but considering the divisive nature of assignment expressions I believe we should eventually _explicitly_ mention their use (either encouraging or discouraging it) in our contributing guide. Thanks for your time :)

**Comment by user27182** [comment]
A few recent PRs I pushed included some uses of the walrus operator (this includes the  cases that were reverted). 

I have no problem avoiding these, and won't use them moving forward.

I would also support removing them from the code base altogether and updating CONTRIBUTING with the policy to not use them.

**Reply by adeak** [reply]
Thanks, and please don't hesitate to disagree with me. Obviously you use them because you see value in them, and that is something I want to respect.

Do you find the syntax more readable? Or is it more about wanting to use concise syntax instead of the more verbose old-style alternatives? (Or something else?)

**Reply by user27182** [reply]
I agree with your suggestion. 

I haven't read the PEP, but my main intent when using it would primarily be to make things more compact (one less line of code), and/or to avoid multiple calls to the same property. But your examples are convincing and show that using the operator makes code hard to read.

So, at best the operator saves one line of code, but at worst it introduces bugs and complicates maintenance. So that makes it an easy decision to just avoid it. It's easy enough to add a line to init a variable explicitly before a bunch of logic instead.

**Reply by adeak** [reply]
Thanks a lot for your thoughts, I much appreciate it.

**Comment by tkoyama010** [comment]
Thanks for pointing this. I would focus on whether the resulting code is readable rather than whether or not to use this expression. For example, I feel stress the definition of this code without walrus operator, which is mentioned in the [Python documentation](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions).
```python
# Before Python3.8, we have to define temporary variable outside of block which is not for me.
n = len(a)
if n > 10:
    print(f"List is too long ({n} elements, expected <= 10)")
```
```python
# After python3.8, we can define inside if block but there is still problem about scope.
if (n := len(a)) > 10:
    print(f"List is too long ({n} elements, expected <= 10)")
```
This may be because I love the idea of scope in C++ and this is close to that representation. Of course, I understand that the walrus operator is problematic in scope as well  which is very unfortunate. What I want from Python is to be able to define scopes explicitly and to be able to temporarily define variables that are valid only inside of those scopes. I am not familiar enough with the Python language to make PEP for it, but I would support such a PEP with all my might.

**Reply by tkoyama010** [reply]
I may have been too focused on scope to properly respond to your point of view. I am all for avoiding the occurrence of spaghetti codes.

**Reply by adeak** [reply]
Thanks for sharing. I guess "use your best judgement" is also possible advice we can give (regarding your "_whether the resulting code is readable_"), but it would also be nice to add some specific examples or counterexamples to guide the reader (because people have wildly warying common senses :smile:).

I'm not sure how I feel about scopes. I suspect it can catch a few rare bugs where a typo (or, more likely, a brain fart) means you use a completely unrelated variable from much earlier in the function. Hmm, and I guess IDEs could make use of the fact that a certain scoped variable is not available later, so tab completion could be more direct (at this point it shows that I don't use an IDE),

**Comment by MatthewFlamm** [comment]
I will reiterate a point made by @tkoyama010 in https://github.com/pyvista/pyvista/discussions/6232#discussioncomment-9721492

> I would focus on whether the resulting code is readable rather than whether or not to use this expression. 

The [Python documentation](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions) also given in that comment says:

> Try to limit use of the walrus operator to clean cases that reduce complexity and improve readability.


I'm marginally in favor of using it in simple examples that more clearly represent the intended scope of use for a variable inside a code block, like a conditional expression, while loop, etc.  I would lean towards discouraging its use otherwise, unless it clearly makes things more readable. I have no strong stake in either case.

**Reply by adeak** [reply]
Thanks for the feedback. I'm also thinking that there are potential use cases, in particular something like a regex match in an `elif` ladder. We sort of have an example of that currently in the codebase, but that specific instance could be avoided (because the branch before that `elif` is `if ...: continue` so the `elif` could be an `if` with a pre-defined variable for the regex match). So even though we don't need this currently, it's plausible to be useful at some point in the future.

**Comment by MatthewFlamm** [comment]
> [The worst example](https://peps.python.org/pep-0572/#simplifying-list-comprehensions):
> 
> ```python
> results = [(x, y, x/y) for x in input_data if (y := f(x)) > 0]
> ```
> 
> This is just tragic. Although it's already a fact in comprehensions that the values are evaluated in the order of
> 
> ```
> [(3. derived value) (1. for variable in iterable) (2. conditions)]
> ```
> 
> the way the definition of the names at the start of the comprehension are bound at the end of the comprehension makes this very hard to read. Just use a loop:
> 
> ```python
> results = []
> for x in input_data:
>     y = f(x)
>     if y > 0:
>         results.append((x, y, x/y))
> ```
> 
> Yes, it's boring, but it's clear.


I would argue that the original sin is the comprehension execution order in this example.  You brought it up several times, but I'd put even more emphasis on it. I am more against the comprehension execution order than the use of walrus operator in this case.  Compare

```py
results = [(x, y, x/y) for x in input_data if (y := f(x)) > 0]
```
 
to

```py
# I'm introducing an arbitrarily chosen colon here to indicate end of conditional expression
# Maybe requirement of a new syntax meaning was what drove the original decision???
results = [for x in input_data if (y := f(x)) > 0 : (x, y, x/y)]  
```

But, this is all discussion in a hypothetical world where both decisions could be made together.  Given that the comprehension syntax is already set in stone and is poor, I am in favor of discouraging adding more complexity to it.  I _think_ this was the original point, or at least close to it, but I added my own color to it.

A more relevant point: We are often working with numpy arrays, so it will be often faster, and more readable, to avoid list comprehensions and use a numpy method instead.  If we are doing numerical operations inside a list comprehension that necessitates a walrus operator, it already smells sub-optimal for many pyvista related operations.

**Reply by adeak** [reply]
> I would argue that the original sin is the comprehension execution order in this example. You brought it up several times, but I'd put even more emphasis on it. I am more against the comprehension execution order than the use of walrus operator in this case.

That's a very good point, thanks. I'm so used to reading it that it doesn't register as something weird. I do like to break comprehensions out into multiple lines:
```python
results = [
    (x, y, x/y)
    for x in input_data
    if (y := f(x)) > 0
]
```
Perhaps I'm especially adverse to assignment expressions in this context because it breaks the rigid pattern that my brain's conditioned to follow when reading a comprehension.

> But, this is all discussion in a hypothetical world where both decisions could be made together. Given that the comprehension syntax is already set in stone and is poor, I am in favor of discouraging adding more complexity to it. I *think* this was the original point, or at least close to it, but I added my own color to it.

I think it was, well put. Although I tend to think that simple list comprehensions are worth using because they make the code _much_ more concise. But this again can't be separated from the fact that I'm too used to using them.
